internal/httpclient/policy_retry_test.go

```diff
@@ -0,0 +1,283 @@
+	}))
+	defer server.Close()
+
+	opts := &RetryOptions{
+		MaxRetries: 2,
+		RetryDelay: 10 * time.Millisecond,
+	}
+	policy := NewRetryPolicy(opts)
+	req, _ := http.NewRequest(http.MethodGet, server.URL, nil)
+
+	next := func(r *http.Request) (*http.Response, error) {
+		return http.DefaultClient.Do(r)
+	}
+
+	resp, err := policy.Do(req, next)
+	if err != nil {
+		t.Fatalf("Expected no error, got: %v", err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	// Should still be 500 after max retries
+	if resp.StatusCode != http.StatusInternalServerError {
+		t.Errorf("Expected status 500 after max retries, got %d", resp.StatusCode)
+	}
+
+	expectedAttempts := 3 // initial + 2 retries
+	if attempts != expectedAttempts {
+		t.Errorf("Expected %d attempts, got %d", expectedAttempts, attempts)
+	}
+}
+
+func TestRetryPolicyCustomStatusCodes(t *testing.T) {
+	attempts := 0
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		attempts++
+		w.WriteHeader(http.StatusBadGateway) // 502
+	}))
+	defer server.Close()
+
+	opts := &RetryOptions{
+		MaxRetries:       2,
+		RetryDelay:       10 * time.Millisecond,
+		RetryStatusCodes: []int{502}, // Only retry on 502
+	}
+	policy := NewRetryPolicy(opts)
+	req, _ := http.NewRequest(http.MethodGet, server.URL, nil)
+
+	next := func(r *http.Request) (*http.Response, error) {
+		return http.DefaultClient.Do(r)
+	}
+
+	resp, err := policy.Do(req, next)
+	if resp != nil {
+		defer func() {
+			_ = resp.Body.Close()
+		}()
+	}
+	
+	if err != nil {
+		t.Fatalf("Expected no error, got: %v", err)
+	}
+
+	expectedAttempts := 3 // initial + 2 retries
+	if attempts != expectedAttempts {
+		t.Errorf("Expected %d attempts for 502 error, got %d", expectedAttempts, attempts)
+	}
+}
+
+func TestRetryPolicyNoRetryOn4xx(t *testing.T) {
+	attempts := 0
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		attempts++
+		w.WriteHeader(http.StatusBadRequest) // 400
+	}))
+	defer server.Close()
+
+	opts := &RetryOptions{
+		MaxRetries: 3,
+		RetryDelay: 10 * time.Millisecond,
+	}
+	policy := NewRetryPolicy(opts)
+	req, _ := http.NewRequest(http.MethodGet, server.URL, nil)
+
+	next := func(r *http.Request) (*http.Response, error) {
+		return http.DefaultClient.Do(r)
+	}
+
+	resp, err := policy.Do(req, next)
+	if err != nil {
+		t.Fatalf("Expected no error, got: %v", err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	// Should not retry on 400
+	if attempts != 1 {
+		t.Errorf("Expected 1 attempt for 400 error (no retry), got %d", attempts)
+	}
+}
+
+func TestRetryPolicyWithLogger(t *testing.T) {
+	attempts := 0
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		attempts++
+		if attempts < 2 {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer server.Close()
+
+	logger := logging.New(logging.DebugLevel)
```

Comment:
Can we test the logger output in this tests ? Making sure we do see the log line. You should be able to use NewWithOutput